File: java/0929.0_Unique_Email_Addresses.java
/*
流式处理

执行用时：76 ms, 在所有 Java 提交中击败了6.45%的用户
内存消耗：41.6 MB, 在所有 Java 提交中击败了61.48%的用户
通过测试用例：185 / 185
*/
class Solution {
    public int numUniqueEmails(String[] emails) {
        return (int) Arrays.stream(emails).map(email -> { 
            email = email.replaceAll("\\+.*@", "@"); 
            while (email.matches(".*\\..*@.*")) { 
                email = email.replaceFirst("\\.", ""); 
            } return email; 
        }).distinct().count();
    } 
}



/*
执行用时：27 ms, 在所有 Java 提交中击败了14.04% 的用户
内存消耗：42.1 MB, 在所有 Java 提交中击败了24.66% 的用户
通过测试用例：185 / 185
*/
class Solution {
    public int numUniqueEmails(String[] emails) {
        Set<String> unique = new HashSet<>();
        for (String email: emails) {
            String[] parts = email.split("@");
            parts[0] = parts[0].split("\\+")[0];
            parts[0] = parts[0].replaceAll("\\.", "");
            unique.add(parts[0] + "@" + parts[1]);
        }
        return unique.size();
    }
}

File: leetcode-cn/0929.0_Unique_Email_Addresses.py
'''
执行用时：44 ms, 在所有 Python3 提交中击败了68.50% 的用户
内存消耗：15.1 MB, 在所有 Python3 提交中击败了65.50% 的用户
通过测试用例：185 / 185
'''
class Solution:
    def numUniqueEmails(self, emails: List[str]) -> int:
        unique = set()
        for email in emails:
            local, domain = email.split('@')
            local = local.split('+')[0]
            local = local.replace('.', '')
            unique.add(local + '@' + domain)
        return len(unique)

'''
执行用时：36 ms, 在所有 Python3 提交中击败了95.25% 的用户
内存消耗：15.1 MB, 在所有 Python3 提交中击败了65.50% 的用户
通过测试用例：185 / 185
'''
class Solution:
    def numUniqueEmails(self, emails: List[str]) -> int:
        unique = set()
        for email in emails:
            local, domain = email.split('@')
            local = local.split('+', 1)[0].replace('.', '')
            unique.add(local + '@' + domain)
        return len(unique)


'''
执行用时：40 ms, 在所有 Python3 提交中击败了85.50% 的用户
内存消耗：14.9 MB, 在所有 Python3 提交中击败了99.00% 的用户
通过测试用例：185 / 185
'''
class Solution:
    def numUniqueEmails(self, emails: List[str]) -> int:
        process = lambda x: x.split('@')[0].split('+', 1)[0].replace('.', '') + '@' + x.split('@')[1]
        return len(set(process(email) for email in emails))


'''
one line
写法：
(lambda function)(arguments)

执行用时：44 ms, 在所有 Python3 提交中击败了68.50% 的用户
内存消耗：15.1 MB, 在所有 Python3 提交中击败了64.00% 的用户
通过测试用例：185 / 185
'''
class Solution:
    def numUniqueEmails(self, emails: List[str]) -> int:
        return len(set((lambda x: x.split('@')[0].split('+', 1)[0].replace('.', '') + '@' + x.split('@')[1])(email) for email in emails))


